### ISSUE TITLE
Porting to ESP-IDF with C++ support

### ISSUE THREAD

**Issue by v1993** [olikraus/u8g2#789]
Hello, I'm interested in porting this great library to ESP32 without Arduino compatiblity layer. However, for me it seems that there is a problem with `cppsrc/U8x8lib.cpp`: it use arduino specific stuff.

Is it possible to add support for other platform in it or it isn't required for C-side calls implementations and headers will be enough (i. e. not build this file at all)?

Is it doing only low-level stuff which I can do in any plain C file, or some things which are reqired to get C++ working too?

**Comment by olikraus** [comment]
U8g2 itself is a pure C library. For Arduino however it calls the C++ Arduino HAL and adds a C++ lib API wrapper. Both, Arduino wrapper and HAL are put together in the same c++/h files. As a consequence: If you want the C++ U8g2 API but not the C++ Arduino HAL, then some code needs to be copied out the existing C++ code. 

I am not 100%, but I think this has been done by the nocemcu people already: https://nodemcu.readthedocs.io/en/dev-esp32/modules/u8g2/

In general porting instructions are here: https://github.com/olikraus/u8g2/wiki/Porting-to-new-MCU-platform

**Comment by v1993** [comment]
Thank you. Now I think about architecture: is it good idea to place ESP32 specific stuff in `sys/esp-idf`? It will be included from `CMakeLists.txt` or `component.mk`. I also plan adding config (for ESP32 menuconfig menu) for 16bit mode and fonts to make u8g2 feel like nice component.

Also, where can I find pure C exmples? I feel like there is a lack of them.

**Comment by olikraus** [comment]
> Now I think about architecture: is it good idea to place ESP32 specific stuff in sys/esp-idf? It will be included from CMakeLists.txt or component.mk

I have no idea about this. CMakeLists.txt or component.mk in the u8g2 are part of a pull request. I can not tell you something about these files.

> Also, where can I find pure C exmples? 

Have a look here: https://github.com/olikraus/u8g2/tree/master/sys
All directories except "arduino" will contain pure c examples.

Here is a "hello world" example for the STM32L031:
https://github.com/olikraus/u8g2/blob/master/sys/arm/stm32l031x6/u8x8_test/main.c

Once I created a blog entry also:
https://drolliblog.wordpress.com/2017/11/02/stm32l031x6-eval-board/

> I feel like there is a lack of them.

I personally feel, that I created more docs and examples than many other Arduino libs...

**Comment by v1993** [comment]
Thanks you againg and sorry for blaming you about no docs. Your Arduino/C++ reference is just great! I just had some problems about pure C documentation.

**Comment by v1993** [comment]
> Thanks you again and sorry for blaming you about no docs. Your Arduino/C++ reference is just great! I just had some problems about pure C documentation.

**Comment by olikraus** [comment]
:-)

Let me know if I can help you with u8g2 related problems.

**Comment by v1993** [comment]
> > Now I think about architecture: is it good idea to place ESP32 specific stuff in sys/esp-idf? It will be included from CMakeLists.txt or component.mk
> 
> I have no idea about this. CMakeLists.txt or component.mk in the u8g2 are part of a pull request. I can not tell you something about these files.

There are no problems with changing those files, just wonder is `sys/platformname` a right place to store platform-specific stuff (I/O frunctions for ESP32 in this case).

**Comment by olikraus** [comment]
> There are no problems with changing those files, just wonder is sys/platformname a right place to store platform-specific stuff (I/O frunctions for ESP32 in this case).

You mean below u8g2? Well sure, it you like to do so and send a corresponding pull request then I can accept this.

**Comment by v1993** [comment]
Thank you! Another question: how presice should be all the `U8X8_MSG_DELAY_*` stuff should be at waiting?

There are three options: use `vTaskDelay` (round time to ticks, pretty rough), `ets_delay_us` for busywait correct time (much better at timing, but not let anything else run while waiting), or `ets_delay_us` with disabeling interrupts with `portDISABLE_INTERRUPTS`/`portENABLE_INTERRUPTS` (100% correct timing, but extremely bad at long and not that presice waitings due to not be able to handle interrupts).

I suggest using `vTaskDelay` on `≥ 10ms` delays and `ets_delay_us` on smaller scale. Are there any parts which should be more accurate at timing (so disable interrupts)? If so, it's still better to disable them in actual calling code and not callback.

Thanks.

**Comment by olikraus** [comment]
> Another question: how presice should be all the U8X8_MSG_DELAY_* stuff should be at waiting?

It always can be more than requested, so: It does not need to be precise.

> Are there any parts which should be more accurate at timing (so disable interrupts)?

I think the only sensitive message is the I2C delay. But this is only required for software emulated I2C.

> Where code generator and template files are?

You probably do not need to take care on the code generator. "codebuild" only connects the displays with the communication interfaces: It defines which types of interfaces are available for which display.

The hardware specific files are either in the sys/<platform>/common directory or (an exception) for  Arduino in the u8g2/cppsrc directory.

> My current idea it to leave csrc as is

Yes please. "csrc" should not contain any hardware specific code.

**Comment by v1993** [comment]
Are Arduino display C++ binding auto generated? If yes, because I want to support them on ESP-IDF too, it will require tweaking code generator.

Am I right?

My goal is to provide C++ bindings compatible with Arduino interface too, because C++ is commonly used on ESP32 too.

**Comment by olikraus** [comment]
> Are Arduino display C++ binding auto generated?

If you mean the constructor classes here, then yes:
https://github.com/olikraus/u8g2/blob/master/cppsrc/U8g2lib.h#L404

**Comment by olikraus** [comment]
Or you create a copy of the codebuild Code and add your modifications there.

**Comment by v1993** [comment]
To be fair, current code generator looks like hell for me ☺.

I'll look at it when get pure C working.

Do you have anything againist re-writing in in something non-C? I'm good (enough to write such thing) with C++ and lua.

**Comment by v1993** [comment]
Can you take a look an my test implementation for `gpio_and_delay`, please?

```
uint8_t u8x8_gpio_and_delay_espidf(u8x8_t *u8x8, uint8_t msg, uint8_t arg_int, U8X8_UNUSED void *arg_ptr)
{
    uint8_t i;
    switch(msg)
    {
    case U8X8_MSG_GPIO_AND_DELAY_INIT:
        for( i = 0; i < U8X8_PIN_CNT; i++ )
            if ( u8x8->pins[i] != U8X8_PIN_NONE )
            {
            bool success;
                if ( i < U8X8_PIN_OUTPUT_CNT )
                {
					success = u8x8_setpinoutput_espidf(u8x8->pins[i]);
                }
                else
                {
					success = u8x8_setpininput_espidf(u8x8->pins[i]);
                }
                if (!success) return 0;
            }

        break;

    case U8X8_MSG_DELAY_NANO:
        ets_delay_us(arg_int==0?0:1);
        break;

    case U8X8_MSG_DELAY_10MICRO:
        /* not used at the moment */
        break;

    case U8X8_MSG_DELAY_100NANO:
        /* not used at the moment */
        break;

    case U8X8_MSG_DELAY_MILLI:
        vTaskDelay(arg_int);
        break;
    case U8X8_MSG_DELAY_I2C:
        /* arg_int is 1 or 4: 100KHz (5us) or 400KHz (1.25us) */
        ets_delay_us(arg_int<=2?5:2);
        break;
    case U8X8_MSG_GPIO_I2C_CLOCK:
    case U8X8_MSG_GPIO_I2C_DATA:
		auto pin = u8x8_GetPinValue(u8x8, msg);
        if ( arg_int == 0 )
        {
            if (
            (!u8x8_setpinoutput_espidf(pin)) ||
            (gpio_set_level(pin, 0) != ESP_OK)
            ) return 0;
        }
        else
        {
            if (!u8x8_setpininput_espidf(pin)) return 0;
        }
        break;
    default:
        if ( msg >= U8X8_MSG_GPIO(0) )
        {
            i = u8x8_GetPinValue(u8x8, msg);
            if ( i != U8X8_PIN_NONE )
            {
                if ( u8x8_GetPinIndex(u8x8, msg) < U8X8_PIN_OUTPUT_CNT )
                {
                    gpio_set_level(i, arg_int);
                }
                else
                {
                    u8x8_SetGPIOResult(u8x8, gpio_get_level(i));
                }
            }
            break;
        }

        return 0;
    }
    return 1;
}
```
Are there any visible flaws? Thanks.

**Comment by olikraus** [comment]
> To be fair, current code generator looks like hell for me 

then, it is better to work on a copy... ;-)

> Do you have anything againist re-writing in in something non-C?

I do not care if you work on your own repo ;-)
For pull requests: New files in new directories should be ok. For the existing code I prefer C over C++

**Comment by v1993** [comment]
Hmm, seems there is a problem…

Arduino introduce global SPI state, but ESP-IDF does not. Doing so would mean big problems with more than displays running in parallel. It would be really nice to have non-interputtable version where SPI state don't have to be stored between calls. Any ideas?

Possible solution is to add SPI and I2C states to `u8x8_t` struct on ESP32. How bad is it? In my opinion, that's ok, but you're main developer after all…

**Comment by v1993** [comment]
BTW, do you have an IRC/Jabber? It probably would be faster to communicate there (nothing personal, feel free to reject with or without reason).

**Comment by v1993** [comment]
Allright, got it! There IS global state on hardware level (`SPI1` and `SPI2`), so I make one "universal" function which can work with any SPI descriptor and call it from `SPI(1/2)` handlers.

**Comment by v1993** [comment]
Well, just got that I was doing HW SPI all wrong… It's better try to get ESP-IDF driver working and NOT try to adapt one from arduino-esp32.

I have some questions now:

1. How big data transfers usually are? There are huge optimisations for < 32 byte ones.
2. Do all SPI stuff have to be syncronized?
3. Do U8G2 read anything? I'm pretty sure it isn't, just want to be 100% clear about it.

**Comment by olikraus** [comment]
> Possible solution is to add SPI and I2C states to u8x8_t struct on ESP32. How bad is it?

Yes, it will break the independence from SPI and I2C

> BTW, do you have an IRC/Jabber?

No

> How big data transfers usually are? There are huge optimisations for < 32 byte ones.

I2C transfers are limited to 32 bytes. Any other transfers can have any length. 

> Do all SPI stuff have to be syncronized?

Against what?

> Do U8G2 read anything? I'm pretty sure it isn't, just want to be 100% clear about it.

U8g2 does not read from displays (but from GPIOs)

Just a general notes: I do not have much time to work on U8G2. So I can not do more than give comments on your questions. 
The core part of u8g2 has to be C and it must not depend on any low level, system specific code... at least if you want me to accept pull requests.

**Comment by v1993** [comment]
Ok, thank you.

> > Do all SPI stuff have to be syncronized?
> 
> Against what?

In this case syncronized means no caching in driver, i.e. use polling (write means waiting until transfer is complete) and not interrupt (things goes into queue and are done in background) SPI mode (ESP-IDF specific stuff).

P.S.: I was talking about HW SPI transfers, or to be presice `arg_int` for `U8X8_MSG_BYTE_SEND` message. Sorry for not being clear.

**Comment by v1993** [comment]
Also, if adding stuff to `u8x8_t` is bad, can I use `U8X8_WITH_USER_PTR` and store my stuff by that pointer?

**Comment by v1993** [comment]
Hi. My porting is going ok yet. There is a question: can I be sure that memory pointed by `arg_ptr` is HW SPI callback is on stack? Thanks.

**Comment by olikraus** [comment]
> Also, if adding stuff to u8x8_t is bad, can I use U8X8_WITH_USER_PTR and store my stuff by that pointer?

Yes, sure, that is the purpose of the user ptr

> There is a question: can I be sure that memory pointed by arg_ptr is HW SPI callback is on stack?

probably not.

**Comment by v1993** [comment]
Interesting thing: it's possible to use two displays with all same pins except CS which is handled by driver (even from two cores)!

However, I want to know one thing: is it possible to trace destruction of u8x8? I know that it's rare thing, but it would be nice to delete device in such a case.

Thanks.

**Comment by v1993** [comment]
And another (more actual) question: can I make few SPI transactions instead of one?

**Comment by olikraus** [comment]
> Interesting thing: it's possible to use two displays with all same pins except CS (and DC I think) which is handled by driver (even from two cores)!

Yes, sure, just create two different u8x8 / u8g2 structures (classes).

> However, I want to know one thing: is it possible to trace destruction of u8x8?

What do you want to trace? U8g2/U8x8 do not use dynamic memory.

> I know that it's rare thing, but it would be nice to delete device in such a case.

There are many nice features. All of them will cost resources. And people already complain about u8g2 size. Nevertheless, i think this is very much possible.

**Comment by v1993** [comment]
At first, what about my last question?

Arduino have SPI API like this:

* Begin transaction
* Write byes
* End transaction

ESP-IDF, on other hand, does it in different way:

1. * Add transaction (complete) into queue
   * Request results later
2. * Do synchronous transaction (complete memory required again)
Under complete memory I mean that only bytes available at moment of beginning transaction will be sent in it.

So, is it ok to perform complete transaction on every "write" and ignore transaction begin/end (CS managed by ESP-IDF SPI driver itself)?

**Comment by olikraus** [comment]
> So, is it ok to perform complete transaction on every "write" and ignore transaction begin/end (CS managed by ESP-IDF SPI driver itself)?

Unknown. It probably will depend on the display controller. It is like this: The display controller will receive commands from u8g2. A command sometimes include a command byte and an argument bytes, so it may consist of two (or more) bytes. A controller may allow the reception of these two bytes in two different transactions, but this is not 100% clear to me. You need to study the graphics controller datasheets for this. In u8g2 it is ensured, that command and args are always sent in the same transaction (the CS line does not change between command byte and argument byte of the same full command).

**Comment by v1993** [comment]
And about "destructor":

U8g2 itself don't use dynamic memory, but some implementations (like mine) may need to or require some hardware descriptors. Possible implemetation is quite simple:

* Add `U8X8_MSG_GPIO_AND_DELAY_CLEAN` and pass it to main callback.
* Let C users call `u8x8_destroy()` or something like that.
* For C++, call it from actual destructor.

Souldn't be hard or breaking, right?



> A command sometimes include a command byte and an argument bytes, so it may consist of two (or more) bytes.

Hmm, seems pretty bad. Is it accepable to just send them as usually or will I have to use [these fields](https://docs.espressif.com/projects/esp-idf/en/latest/api-reference/peripherals/spi_master.html#_CPPv2N17spi_transaction_t3cmdE)? That would be complicated.

Another solution is to allocate buffer for SPI stuff and do actual transaction on "end transaction".

> You need to study the graphics controller datasheets for this.

I'm not ready to do so on all displays supported by u8g2…

Thansk.

**Comment by v1993** [comment]
And another question: what is max size of SPI transaction? I have to set in bus configuration.

Can I rely on default 4094 bytes?

**Comment by olikraus** [comment]
> Can I rely on default 4094 bytes?

It will probably much shorter. For I2C it is limited to 32 for SPI I assume it will not exceed 512 bytes. The only exception might be e-paper devices, which might require larger blocks.

**Comment by v1993** [comment]
Ok, thanks. I'll add into config then (with 512 as default).

**Comment by v1993** [comment]
Should CS line be positive or negative during communication?

Default driver's behaviour is to set it ti negative, but there is a flag `SPI_DEVICE_POSITIVE_CS`. Should I use it?

P.S.: what about destructor idea? Thanks.

**Comment by v1993** [comment]
> > And yet another question: can SPI mode change for device?
> 
> Yes of course.

I mean, for one device between different transactions? If still yes, that's pretty bad, because it mean that I have to remove and add device again, reducing speed a lot.

> > P.S.: what about destructor idea? Thanks.
> 
> U8g2 is fine with the C++ default constructor, because there is no further dynamic memory allocation.

As I already said, implementation may need to do internal cleanup. I have to `malloc` things to store custom stuff.

> > Should CS line be positive or negative during communication?
> 
> Depends on the display:
> [/csrc/u8x8.h@`master`#L224](https://github.com/olikraus/u8g2/blob/master/csrc/u8x8.h#L224)

If `chip_enable_level == 1`, should it be negative or positive during activity?

**Comment by olikraus** [comment]
> > > And yet another question: can SPI mode change for device?
> > 
> > Yes of course.
> I mean, for one device between different transactions? If still yes, that's pretty bad, because it mean > that I have to remove and add device again, reducing speed a lot.

no

> > > P.S.: what about destructor idea? Thanks.
> > 
> > U8g2 is fine with the C++ default constructor, because there is no further dynamic memory allocation.
> As I already said, implementation may need to do internal cleanup. I have to `malloc` things to store > custom stuff.

I assume you will create a new interface, so of course, you should provide a proper constructor.

> > > Should CS line be positive or negative during communication?
> > 
> > Depends on the display:
> > [/csrc/u8x8.h@`master`#L224](https://github.com/olikraus/u8g2/blob/master/csrc/u8x8.h#L224)
> If `chip_enable_level == 1`, should it be negative or positive during activity?

`chip_enable_level` is not a variable, it is a calibration value which returns the level. If your transfer is started, you should set the chip select level like this:
```
  case U8X8_MSG_BYTE_START_TRANSFER:
      u8x8_gpio_SetCS(u8x8, u8x8->display_info->chip_enable_level);  
```

> Can say one thing for sure: SW I2C for SSD1306 from pure C is working just fine!

nice

**Comment by v1993** [comment]
I'm having a strange problems when try to get HW I2C working.

Display correctly responds with ACK on data and adress, but nothing works. If I try rebooting from SW to HW implementation (without changing screen), some black lines appears, like this:

![img_20190120_204730_1](https://user-images.githubusercontent.com/16255906/51443166-b6be4980-1cf5-11e9-9e9c-cec588497016.jpg)

Can you guess any reason? Thanks.

**Comment by olikraus** [comment]
Difficult to say. Speed might be to high. Or the picture just reflects the content from the last correct update with some garbage from the restart.

**Comment by v1993** [comment]
At first, I started SW I2C version which draw the picture. Then flased HW I2C version, then rebooted few times after finishing (more lines after each restart).

Display changes only when actual sending occours (I use logging to PC terminal).

I tried changing speed, but can't get working display this way.

Tried speeds:

* 400000Hz (from U8G2)
* 100000Hz
* 10000Hz
* 1000Hz

Any other suggestions?

**Comment by olikraus** [comment]
Your low level code might be wrong

**Comment by v1993** [comment]
Hmm, I have a theory: probably `i2c_master_write(u8x8->user_ptr, arg_ptr, arg_int, true);` don't copy memory but only store pointer? In this case it may point to invalid memory at time of transaction.

Only a theory…

**Comment by olikraus** [comment]
user_ptr just is a pointer value, which you provide. It does not copy anything (not sure if this helps)

**Comment by v1993** [comment]
Yep, I know. I store transaction pointer there.

Just checked source code of that `i2c_master_write`, looks like it doesn't copy memory. So I really probably am right…

Is there any limit on how many writes may go in one I2C transaction?

**Comment by v1993** [comment]
Got HW I2C working :smile:

I copy incoming data into dynamically mallocated memory and then free it after sending. It have pretty big perfomance impact.

Are I2C transactions limited to some size? Not "write" calls to I2C callback, but I2C transactions themselves. This way it would be possible to allocate buffer only once.

Also, how many times per transaction "write" callback can be called?

Thank you.

**Comment by v1993** [comment]
Aaand my luck just ran out: SPI capable diplay isn't working :(

I'll have to order another one and wait long time before will be able to test it.

Are there ESP32 and SPI users?

**Comment by v1993** [comment]
Hi, @olikraus, it's me again. Can you, please, provide some hints on codebuild's structure and functions?  I want to try to re-write it in a bit more readable manner.

Thank you.

**Comment by olikraus** [comment]
what exactly do you want to know?

**Comment by v1993** [comment]
What files are generated by it?

What different structures does?

**Comment by olikraus** [comment]
> What files are generated by it?

The following files are generated:
u8g2_d_memory.c
u8g2_d_setup.c
The following files are modified:
u8g2.h
U8g2lib.h
U8x8lib.h
u8g2setupc.md
u8g2setupcpp.md
u8x8setupc.md
u8x8setupcpp.md

> What different structures does?

struct interface: Describes the communication interfaces like SPI etc
struct controller: Describes the controller and the supported displays for this controller

**Comment by olikraus** [comment]
i think this can be closed

**Comment by v1993** [comment]
Hi @olikraus, I'm back! :smiley:
Finally recivied a working SPI display and got some time to expirement with it. I'm getting weird glitched image and have no idea what could be reason behind it (transaction content itself seems fine). Can you take a look at this video, please? https://youtu.be/Vzd91BQ8rZw

I run following things every frame:

    u8g2.setDrawColor(1);
    u8g2.drawBox(0, 0, 255, 127);
    u8g2.sendBuffer();

Similliar results when I draw some text.

Program works 100% percent correctly when I use I²C. Do you have an idea what could be happening? I don't think that it is garbage memory because I've double-checked things there with hex dumping and everything seems fine.

**Comment by olikraus** [comment]
Hmmm... Did you enable 16 bit mode? See FAQ.

**Comment by v1993** [comment]
I'm pretty sure I did. Also, why would it work in I²C and fail in SPI because of 8bit mode anyways?

**Comment by v1993** [comment]
I think that it is most probably some wrong setting for SPI driver.

As a sidenote, is it ok to let driver handle CS line in auto mode? ESP-32 SPI driver is capable of managing it and I use such ability.

Here's my driver code for reference:

```C++
static uint8_t u8x8_byte_espidf_hw_spi_universal(u8x8_t* u8x8, uint8_t msg, uint8_t arg_int, void* arg_ptr, uint8_t host)
// It can work with any of two SPIs
	{
//#ifdef U8X8_HAVE_HW_SPI
	u8x8_device_info* info = u8x8->user_ptr;
	switch (msg) {
			case U8X8_MSG_BYTE_SEND: ;

				// We add stuff into buffer and then send it in one transaction
				//if (info->spi.bufsiz != 0)
				//	hexDump("Old buf", info->spi.buffer, info->spi.bufsiz);
				size_t oldsize = info->spi.bufsiz;
				info->spi.bufsiz += arg_int;
				//hexDump("Appending", arg_ptr, arg_int);
				info->spi.buffer = heap_caps_realloc(info->spi.buffer, info->spi.bufsiz, MALLOC_CAP_DMA); // FIXME: MALLOC_CAP_32BIT may crash stuff
				memcpy(info->spi.buffer + oldsize, arg_ptr, arg_int);
				//hexDump("New buf", info->spi.buffer, info->spi.bufsiz);

				break;
			case U8X8_MSG_BYTE_INIT: ;
				info->type = SPI;
				info->spi.device = NULL;
				info->spi.buffer = NULL;
				info->spi.bufsiz = 0;
				spi_bus_config_t busconf = {};
				busconf.mosi_io_num = u8x8->pins[U8X8_PIN_SPI_DATA];
				busconf.miso_io_num = -1;
				busconf.sclk_io_num = u8x8->pins[U8X8_PIN_SPI_CLOCK];
				busconf.quadwp_io_num = -1;
				busconf.quadhd_io_num = -1;
				busconf.max_transfer_sz = 0; // FIXME: add it into config
				busconf.flags = 0; // FIXME: add stuff here?
				if (u8x8->bus_clock == 0) 	/* issue 769 */
					u8x8->bus_clock = u8x8->display_info->sck_clock_hz;
				/* disable chipselect */
				//u8x8_gpio_SetCS(u8x8, u8x8->display_info->chip_disable_level);

				ESP_ERROR_CHECK(spi_bus_initialize(u8x8_spibuses[host], &busconf, host + 1)); // NOTE: user can set up it manually

				spi_device_interface_config_t conf = {};
				conf.command_bits = 0;
				conf.address_bits = 0;
				conf.dummy_bits = 0; // FIXME: Hmmm, may cause problems?
				conf.mode = u8x8->display_info->spi_mode;
				conf.duty_cycle_pos = 0; // I'm not sure
				conf.clock_speed_hz = u8x8->bus_clock; // TODO: add setting to overclock displays :)
				conf.input_delay_ns = 0;
				conf.spics_io_num = u8x8_GetPinIndex(u8x8, U8X8_MSG_GPIO_CS);
				conf.flags = SPI_DEVICE_POSITIVE_CS | SPI_DEVICE_NO_DUMMY;
				if (u8x8->display_info->chip_enable_level) conf.flags |= SPI_DEVICE_POSITIVE_CS; // FIXME: may be right in complete reverse
				conf.queue_size = 1; // I don't quite get it, but ok…
				spi_bus_add_device(u8x8_spibuses[host], &conf, &(info->spi.device)); // FIXME: remove in destructor

				break;

			case U8X8_MSG_BYTE_SET_DC:
				u8x8_gpio_SetDC(u8x8, arg_int);
				break;

			case U8X8_MSG_BYTE_START_TRANSFER:

				// Nothing to do here
				break;

			case U8X8_MSG_BYTE_END_TRANSFER: ;
				// Docs says that it's better to have 32-bit aligned memory
				// Doing it here should make stuff faster
				//info->spi.buffer = heap_caps_realloc(info->spi.buffer, info->spi.bufsiz, MALLOC_CAP_DMA);
				spi_transaction_t transaction = {};
				transaction.flags = 0;
				transaction.length = (info->spi.bufsiz) * 8; // Bits
				transaction.tx_buffer = info->spi.buffer;
				transaction.rx_buffer = NULL; // U8g2 don't care about feedback, bad guys!
				spi_device_polling_transmit(info->spi.device, &transaction);
				free(info->spi.buffer);
				info->spi.buffer = NULL;
				info->spi.bufsiz = 0;

				break;
			default:
				return 0;
			}

	return 1;
	}
```
P.S.: can you add destructor event, please? Something like `U8X8_MSG_DESTROY` passed to GPIO callback, being ignored by anyone who don't have to clear stuff and handled by ones who do. I really need it!

**Comment by olikraus** [comment]
> P.S.: can you add destructor event, please?

There is already a default destructor: https://github.com/olikraus/u8g2/issues/815

> Here's my driver code for reference:

Maybe there is something wrong in your SPI code. At least it looks complicated to me.

**Comment by v1993** [comment]
>There is already a default destructor: #815

That destructor story is TOTALLY different. That person just had no understanding of basic C++ concepts, including fact that destructors are called automatically by language. I, on other hand, what to have C-side destructor (or cleanup function which is same thing called the other way) which could be called eiter manually (if using C interface, like `u8x8_destroy(u8x8);`) or automatically by C++ (declare C++ destructor which calls C one).

> At least it looks complicated to me.

Well, most of code is just creating various structures if you look closely. The rest is maintaining simple buffer which gets extended/allocated by `heap_caps_realloc`, which is pretty much same as `realloc` except for last argument.

**Comment by v1993** [comment]
Just figured out faulty string:

    conf.spics_io_num = u8x8_GetPinIndex(u8x8, U8X8_MSG_GPIO_CS);
Replaced it with

    conf.spics_io_num = u8x8_GetPinValue(u8x8, U8X8_MSG_GPIO_CS);
Still face the same issue tho :\

**Comment by olikraus** [comment]
> That destructor story is TOTALLY different. 

Well, I can only say this for u8g2: If you want me to create a destructor, then the question is what shell I do inside the u8g2 destructor? Or in other words: How should the destructor differ from the default destructor which is already there.

> Well, most of code is just creating various structures if you look closely.

My motivation is not that high, to look into this closely. I mean, my own code works correctly. Your code makes use of many  other functions, which I do not know. And: I already looked into it closely: For example you call `heap_caps_realloc` but then it is released with `free`. Is this correct? I do not know this, at least it sounds strange.

**Comment by v1993** [comment]
As for destructor thing, it's probably easier to implement and make pull request rather than describe (yeah, I sometimes suck at describing :disappointed:).

> And: I already looked into it closely: For example you call `heap_caps_realloc` but then it is released with `free`. Is this correct? I do not know this, at least it looks strange.

It is correct. According to [ESP-IDF docs](https://docs.espressif.com/projects/esp-idf/en/latest/api-reference/system/mem_alloc.html#_CPPv414heap_caps_freePv):

> `void heap_caps_free(void *ptr)`
> Free memory previously allocated via `heap_caps_malloc()` or `heap_caps_realloc()`.
> …
> In IDF, `free(p)` is equivalent to `heap_caps_free(p)`.

But it's probably better ro replace it with caps version for better readability and potentially better compatiblity with futher IDF versions.

ESP-IDS SPI driver use 32bit blocks and say to handle non-32bit ones in some way unclear for me. When I do so, however, display doesn't show anything at all. Aligning memory to 32bits do nothing too.

P.S.: can you reopen this, please? I don't feel that comfortable discussing problems in closed issue :).

**Comment by v1993** [comment]
Here's my code after some minor fixes and cleanup:

```C++
static const spi_host_device_t u8x8_spibuses[2] = {HSPI_HOST, VSPI_HOST};

static uint8_t u8x8_byte_espidf_hw_spi_universal(u8x8_t* u8x8, uint8_t msg, uint8_t arg_int, void* arg_ptr, uint8_t host)
// It can work with any of two SPIs
	{
	u8x8_device_info* info = u8x8->user_ptr;
	switch (msg) {
			case U8X8_MSG_BYTE_SEND:
				;

				// We add stuff into buffer and then send it in one transaction
				//if (info->spi.bufsiz != 0)
				//	hexDump("Old buf", info->spi.buffer, info->spi.bufsiz);
				size_t oldsize = info->spi.bufsiz;
				info->spi.bufsiz += arg_int;
				//hexDump("Appending", arg_ptr, arg_int);
				info->spi.buffer = heap_caps_realloc(info->spi.buffer, info->spi.bufsiz, MALLOC_CAP_DMA);
				memcpy(info->spi.buffer + oldsize, arg_ptr, arg_int);
				//hexDump("New buf", info->spi.buffer, info->spi.bufsiz);

				break;
			case U8X8_MSG_BYTE_INIT:
				;
				info->type = SPI;
				info->spi.device = NULL;
				info->spi.buffer = NULL;
				info->spi.bufsiz = 0;
				spi_bus_config_t busconf = {};
				busconf.mosi_io_num = u8x8->pins[U8X8_PIN_SPI_DATA];
				busconf.miso_io_num = -1;
				busconf.sclk_io_num = u8x8->pins[U8X8_PIN_SPI_CLOCK];
				busconf.quadwp_io_num = -1;
				busconf.quadhd_io_num = -1;
				busconf.max_transfer_sz = 0; // FIXME: add it into config
				busconf.flags = 0; // FIXME: add stuff here?
				if (u8x8->bus_clock == 0) {	/* issue 769 */
						u8x8->bus_clock = u8x8->display_info->sck_clock_hz;
						}
				/* disable chipselect */
				//u8x8_gpio_SetCS(u8x8, u8x8->display_info->chip_disable_level);

				spi_bus_initialize(u8x8_spibuses[host], &busconf, host + 1);    // NOTE: user can set up it manually

				spi_device_interface_config_t conf = {};
				conf.command_bits = 0;
				conf.address_bits = 0;
				conf.dummy_bits = 0; // FIXME: Hmmm, may cause problems?
				conf.mode = u8x8->display_info->spi_mode;
				conf.duty_cycle_pos = 0; // I'm not sure
				conf.clock_speed_hz = u8x8->bus_clock;
				conf.input_delay_ns = 0;
				conf.spics_io_num = u8x8->pins[U8X8_PIN_CS];
				conf.flags = SPI_DEVICE_NO_DUMMY;
				if (u8x8->display_info->chip_enable_level) {
						conf.flags |= SPI_DEVICE_POSITIVE_CS;
						}
				conf.queue_size = 1; // I only use polling transactions, but ok
				ESP_ERROR_CHECK(spi_bus_add_device(u8x8_spibuses[host], &conf, & (info->spi.device)));         // FIXME: remove in destructor

				break;

			case U8X8_MSG_BYTE_SET_DC:
				u8x8_gpio_SetDC(u8x8, arg_int);
				break;

			case U8X8_MSG_BYTE_START_TRANSFER:

				// Nothing to do here
				break;

			case U8X8_MSG_BYTE_END_TRANSFER:
				;
				// Docs says that it's better to have 32-bit aligned memory
				// Doing it here should make stuff faster
				info->spi.buffer = heap_caps_realloc(info->spi.buffer, info->spi.bufsiz, MALLOC_CAP_DMA | MALLOC_CAP_32BIT);
				/*
				size_t incomplete_len = info->spi.bufsiz % 4U;
				if (incomplete_len > 0) {
					uint32_t* lastbyte = (uint32_t*)(info->spi.buffer - incomplete_len);
					*lastbyte = SPI_SWAP_DATA_TX(*lastbyte, incomplete_len*8);
					ESP_LOGI("u8x8", "Hi there! Size: %u, mistake: %u", info->spi.bufsiz, incomplete_len);
				}
				*/
				spi_transaction_t transaction = {};
				transaction.flags = 0;
				transaction.length = (info->spi.bufsiz) * 8;   // Bits
				transaction.tx_buffer = info->spi.buffer;
				transaction.rx_buffer = NULL; // U8g2 don't care about feedback, bad guys!
				spi_device_polling_transmit(info->spi.device, &transaction);
				heap_caps_free(info->spi.buffer);
				info->spi.buffer = NULL;
				info->spi.bufsiz = 0;

				break;
			default:
				return 0;
			}

	return 1;
	}
```
It still have the same problem…

Can you see any possible misuse of u8g2 stuff here?

**Comment by v1993** [comment]
I've recorded longer video of problem: https://youtu.be/K7tfP_Z6994. It shows strange pattern: blocks move to the right and go one block down with wrapping to upper edge each time they reach right side of screen.

**Comment by olikraus** [comment]
It looks ok to me, except for the handling of the CS line.
Some other remarks:
 * Assignment `conf.flags = SPI_DEVICE_NO_DUMMY;` will be overwritten late with 0
 * `conf.mode = u8x8->display_info->spi_mode;` might be wrong, because the u8g2 SPI mode is the Arduino SPI mode, which might be something completly different compared to the ESP32 SPI mode.

**Comment by v1993** [comment]
> * Assignment `conf.flags = SPI_DEVICE_NO_DUMMY;` will be overwritten late with 0

Where it does? `busconf` is different from `conf` if you are talking about that.

> * `conf.mode = u8x8->display_info->spi_mode;` might be wrong, because the u8g2 SPI mode is the Arduino SPI mode, which might be something completly different compared to the ESP32 SPI mode.

Arduino refer to wikipedia for SPI modes information. ESP32 seems standard-compliant too. I'll expirement with hadcoding settings and checking out results, but not sure will it gite anything.

As for CS line handling, I give it to driver which is capable of managing few devices on same bus (and setting CS to correct values during transactions). So it is fine.

**Comment by v1993** [comment]
Tried changing SPI modes. Display glitches the same way in 0,2,3'th ones and don't work at all in 1'st one (counting from zero).

**Comment by v1993** [comment]
I've been logging canges and got this:

```
D (40826) U8G2: SPI transaction start (formally)
D (40836) U8G2: DC value: 0
D (40836) U8G2: Appending buffer
D (40836) U8G2: DC value: 0
D (40846) U8G2: Appending buffer
D (40846) U8G2: DC value: 0
D (40846) U8G2: Appending buffer
D (40856) U8G2: DC value: 0
D (40856) U8G2: Appending buffer
D (40856) U8G2: DC value: 1
D (40866) U8G2: Appending buffer
D (40866) U8G2: Executing SPI transaction of size 132 bytes
```

So there's how it works on arduino:

1. Start transaction
2. Enable CS
3. Disable DC
4. Send some bytes
5. Enable DC
6. Send more bytes
7. Disable CS
8. End transaction

And here's how it works on ESP-32:

1. Disable DC
2. Enable DC
3. Execute transaction (auto CS management)

Is DC timing during data transfer critical? If so, I have to send buffer every time DC changes I guess.

**Comment by v1993** [comment]
Aha, got it! I have to end data each DC changes. Now display is functioning correctly and give 173 FPS. (!)

**Comment by v1993** [comment]
Got 321FPS on nearly max optimisation. That's just crazy :rofl:.

**Comment by v1993** [comment]
> For I2C it is limited to 32 for SPI I assume it will not exceed 512 bytes. The only exception might be e-paper devices, which might require larger blocks.

You were wrong there. Most SPI transactions (from start to end) use 1052 bytes for my SSD1306.

**Comment by olikraus** [comment]
> Where it does? busconf is different from conf if you are talking about that.

Ah, yes, true. As I said, I shouldn't review unknown data structures.

> Is DC timing critical during data transfer? If so, I have to send buffer every time DC changes I guess.

Indeed, this is the problem. DC is the 9th bit for each byte. It must be set/cleared with each byte. This means you can only collect bytes as long as the DC value is the same. If the DC value changes, you have to flush the "so far collected" bytes, change the DC level and restart to collect the bytes.

DC means Data / Command and is a hint for the controller how to interpret the byte (as data or instruction).

**Comment by v1993** [comment]
As I said, display is working now. I'm busy polishing SPI diver and some other stuff.

What changed in u8g2 since January? Which changes should be ported into ESP-32 versions of `U8x8lib.hpp` and `U8g2lib.hpp`?

After completing work on SPI I'm going to take a look at code generator. Is it ok if I'll redo it not into C but Lua or at least C++?

**Comment by olikraus** [comment]
> What changed in u8g2 since January? Which changes should be ported into ESP-32 versions of U8x8lib.hpp and U8g2lib.hpp?

Good question. The release documentation should be complete, also git history should be there.

> After completing work on SPI I'm going to take a look at code generator. Is it ok if I'll redo it not into C but Lua or at least C++?

U8g2 is open source.... But I personally will keep on adding new devices to the u8g2 c code.

**Comment by v1993** [comment]
> U8g2 is open source.... But I personally will keep on adding new devices to the u8g2 c code.

Well, I actually plan to merge my fork into mainstream, so it is better to keep one generator… ESP-IDF support C++ too, so it is required to tweak generator and make it more flexible. It sould not be too hard to move exsisting either into lua or C++.

**Comment by olikraus** [comment]
Well, yes

**Comment by v1993** [comment]
So it **is** ok to move onto lua or C++?

**Comment by olikraus** [comment]
Depends. Basically I do want additional effort with it.
The existing c code should be compileable under C++. We can add same extra #defines for you if required.

**Comment by v1993** [comment]
Why there is different order of arguments for SW I²C and HW I²C? Is it ok to leave only one function in my port (because they do basicly the same thing)?

**Comment by olikraus** [comment]
> Why there is different order of arguments for SW I²C and HW I²C? 

The sequence was driven by the signal importance the question whether a signal can "reset" was at the end). For hardware SPI in 99% of all cases, the clock and data lines are predefined, so I removed them from the start. Later ESP users asked be to implement the hardware pin remapping for the I2C pins, which, again is optional. So I added this to the end in order to stay compatible with the existing hardware SPI interface.

> Is it ok to leave only one function in my port (because they do basicly the same thing)?

sure... as long as u8g2 is not affected....

> because they do basicly the same thing

SW SPI ignores the I2C ACK response. There are some SSD1306 displays out there, which do not send the I2C ACK. For these displays only the SW I2C will work.

**Comment by v1993** [comment]
> SW SPI ignores the I2C ACK response. There are some SSD1306 displays out there, which do not send the I2C ACK. For these displays only the SW I2C will work.

You meant I²C, right? I was talking about pin configuration functions, but thank you for info! There's an option to don't wait for ACK in ESP-IDF I²C driver. How should I add support for it? I currently see few ways:

1. Add it into ESP-IDF's config menu. Sounds like a bad idea for me because of being outside of code and applied to any used display.
2. Make it custom function only for ESP-32. Could be not perfect but ok overall.
3. Make it standard function for u8g2. Not sure is is a good idea considering fact that arduino don't support it at all.

What do you think?

P.S.: looks like 642.857143 FPS (!!!) is the most I can get out of my display. Very impressive mor me tho!

**Comment by olikraus** [comment]
> You meant I²C, right? 

Ah yes, fixed.

> P.S.: looks like 642.857143 FPS (!!!)

wow...

> What do you think?

I somehow think, that you spent too much effort into it. I think SW I2C is not really required any more. On Arduino (where HW I2C is fixed to specific pins) it makes sense for those users who have to use different pins.

**Comment by v1993** [comment]
@olikraus Well, you say that this is a case when SW I²C is required (when no ACK from display). ESP-IDF driver allow to bypass this problem when using HW I²C. I just aksed what would be a good way to add such an option (for turning ACK wait off).

**Comment by v1993** [comment]
Another very important question: what is the difference between `COM_3WSPI` and `COM_ST7920SPI`? Aren't they basicly the same?

**Comment by olikraus** [comment]
> @olikraus Well, you say that this is a case when SW I²C is required (when no ACK from display). ESP-IDF driver allow to bypass this problem when using HW I²C. I just aksed what would be a good way to add such an option (for turning ACK wait off).

I just wanted to explain the differences between SW I2C and HW I2C. 

> Another very important question: what is the difference between COM_3WSPI and COM_ST7920SPI? Aren't they basicly the same?

No, it differs in the way how one byte of date/command and the dc information is transfered.

**Comment by v1993** [comment]
@olikraus Well, actual thing is HUGE difference in speeds between SW and HW I²C. As you mentioned, there is another one which may lead to using very slow SW version when only one small change to executing code is required to make HW don't wait for ACK (as for ESP-IDF driver). I just want to add option to disable such wait. Sounds like a good idea, right?

> No, it differs in the way how one byte of date/command and the dc information is transfered.

Then

    #define COM_ST7920SPI       0x0020                  /* mostly identical to COM_4WSPI, but does not use DC */
is very misleading. 3 wire SPI don't have DC too if I got everything right way.

Also, what does COM_ST7920SPI flag affect? I can't find any mentions of it inside source code.

P.S.: I'm porting exsisting display data to lua format. Does this look ok?

```lua
…
        {
                name = "ra8835";
                w = 40; h = 30;
                hvline = "u8g2_ll_hvline_horizontal_right_lsb";
                cad = "u8x8_cad_100";
                note = "not tested";
                genclass = true;
                com = mask {"COM_6800";"COM_8080";};
                displays = {
                        "320x240";
                };
        };
        {
                name = "max7219";
                w = 8; h = 1;
                hvline = "u8g2_ll_hvline_horizontal_right_lsb";
                cad = "u8x8_cad_empty";
                genclass = true;
                com = mask {"COM_4WSPI";};
                displays = {
                        "64x8";
                };
        };
…
```

**Comment by v1993** [comment]
`codebuild.c`:

```C
  /* 4 */
  {
    "3W_SW_SPI",
    "u8x8_SetPin_3Wire_SW_SPI",
    "u8x8_byte_arduino_3wire_sw_spi",
    "u8x8_gpio_and_delay_arduino",
    "uint8_t clock, uint8_t data, uint8_t cs, uint8_t reset = U8X8_PIN_NONE",
    "clock, data, cs, reset",
    "clock, data, cs [, reset]",
    "u8x8_byte_3wire_sw_spi"
  },
  /* 8 */
  {
    /* ST7920 */ "SW_SPI",
    "u8x8_SetPin_3Wire_SW_SPI",		/* use the 3 wire interface, because the DC is not used */
    "u8x8_byte_arduino_4wire_sw_spi",	/* improved version over u8x8_byte_4wire_sw_spi */
    "u8x8_gpio_and_delay_arduino",
    "uint8_t clock, uint8_t data, uint8_t cs, uint8_t reset = U8X8_PIN_NONE",
    "clock, data, cs, reset",
    "clock, data, cs [, reset]",
    "u8x8_byte_4wire_sw_spi", /* "u8x8_byte_st7920_sw_spi" */
  },
```

Now I'm pretty lost. My HW SPI driver can handle both DC and non-DC (`dc = U8X8_PIN_NONE`) SPI connections. If DC is not set, I just do nothing on `U8X8_MSG_BYTE_SET_DC`:

https://github.com/v1993/u8g2/blob/1fbbe30384e3fa1534c1aa58bad831c2312af8ee/sys/esp-idf/U8x8lib.c#L309-L315

What type of behavior is this? Is it 3 wire SPI-like, ST7920 SPI-like or some senseless behaviour? What is the difference if there is no DC line at all and all we have is data?

I would be glad to make correctly working HW versions for both modes.

**Comment by olikraus** [comment]
If a controller requires the DC line (like the ST7920 or the SSD1306), the tne DC information has to be transmitted.
This transmission has to happen somehow. It may (!) happen though an extra DC line (and pin).

Speaking about serial communication, there are several methods to transfer the DC information:
I2C: DC infromation is encoded in the byte stream
4W-SPI: DC is part of an extra data line including the DC pin
3W-SPI: DC is encoded into the SPI bit stream: Actually there is no byte stream any more, instead 9 bits per data item are transfered
ST7920 SPI: DC is encoded into the SPI byte stream

> Also, what does COM_ST7920SPI flag affect?

It will use a different CAD procedure (`u8x8_cad_st7920_spi`), but use the classic 4-wire SPI low level proc (altough DC is not required):
```
  {
    "st7920", 	24, 	4, 	"u8g2_ll_hvline_horizontal_right_lsb", "u8x8_cad_st7920_spi", "s", COM_ST7920SPI,
    "", /* is_generate_u8g2_class= */ 1,
    {
      { "192x32" },
      { NULL }
    }
  },
```

**Comment by olikraus** [comment]
Let me say it like this: There are a lot of differences between displays and uC Boards. In order to deal with these differences, I have introduced three callback procedures (layers) in U8g2:
- Protocol layer (CAD procedures in u8g2/u8x8, abbreviation for Command Argument Data): Describes the way how data and commands are convered into bytes on the next lower level (byte transfer layer)
- Byte layer: Transfers 8  data bits and the DC information if the DC is not handled on the protocol layer. 
- Bit layer: gpio and delay procedure

Whether the DC information is handled in protocol layer or in byte layer was a difficult decision and I also changed this a during development. Now, DC is mostly handled in byte layer if there is no DC pin, with the only exception of the 3W-SPI interface.

**Comment by v1993** [comment]
Why there is no HW 3-wire SPI? That shouldn't be too hard, or is it?

If I'm reading this right way:

* 4W-SPI: standard SPI protocol with extra DC line
* 3W-SPI: some improvisation over SPI? Do it still follow SPI protocol rules anyways? Don't have DC line.
* ST7920-SPI: normal SPI, I send exact data I get from CAD. Also, no DC line. In fact, same as 4W-SPI but without DC line.

Do interface flags affect anything expect docs?

Also, what about UART bus? ESP-32 do have one (HW driven) for custom needs. Is there any documentation on how to create u8g2 driver for UART?

**Comment by olikraus** [comment]
> Why there is no HW 3-wire SPI? 

3-wire SPI means data items of 9 bit length. This is not supported by Arduino Lib. 

> 3W-SPI: some improvisation over SPI? Do it still follow SPI protocol rules anyways? 

yes

> Don't have DC line.

no, but the data will carry the DC information in the 9 bit data packet. I think you should study the datasheets for example of the SSD1306.

> ST7920-SPI: normal SPI, I send exact data I get from CAD. Also, no DC line. In fact, same as 4W-SPI but without DC line.

Yes, you can say so. The DC information is encoded also in the SPI bit stream, however the data items are still 8 bits in size.

> Do interface flags affect anything expect docs?

which interface flags?

> Also, what about UART bus? ESP-32 do have one (HW driven) for custom needs. Is there any documentation on how to create u8g2 driver for UART?

The communication in u8g2 is driven by the requirements of the connected displays. There is no display controller which expects RS232.

**Comment by v1993** [comment]
Uneder "interface flags" I mean `COM_ST7920SPI | COM_I2C` stuff.

> The communication in u8g2 is driven by the requirements of the connected displays. There is no display controller which expects RS232.

Huh? Why is there `COM_UART` and what does it mean then?

**Comment by olikraus** [comment]
> Uneder "interface flags" I mean COM_ST7920SPI | COM_I2C stuff.

This is for the code generation if the related constructors

> Huh? Why is there COM_UART and what does it mean then?

This is the exception from the common rule. UART is required for the thermal printer (which is not a display controller ;-)

**Comment by v1993** [comment]
I'm in progress of re-writing codebuild in lua. C part is done :smile:.

**Comment by olikraus** [comment]
sounds good.... still i am not sure what it means for me... :-p

**Comment by v1993** [comment]
BTW that's how I store controller info now: https://github.com/v1993/u8g2/blob/master/tools/newcodebuild/resources/Common/controllers.lua. Doesn't look too complicated, right?

**Comment by v1993** [comment]
As a sidenote, I would be glad to provide 3-wire HW SPI. Can you, please, describe what should this function do for every incoming message?

This is not critical but would be a nice feature overall.

**Comment by olikraus** [comment]
> As a sidenote, I would be glad to provide 3-wire HW SPI. Can you, please, describe what should this function do for every incoming message?

Exactly the same as the software implementation for 3-wire SPI: https://github.com/olikraus/u8g2/blob/master/csrc/u8x8_byte.c#L268

Of course the `U8X8_MSG_BYTE_SEND` will look differently (transmission via the HAL procedures of your uC) and of course you may need to do extra HW SPI init and shutdown during start end end transfer messages (and maybe also during init msg).

**Comment by v1993** [comment]
@olikraus Taking a closer look at that fuction, would that be possible to make 3-wire 
HW SPI just a wrapper around 4-wire one?

I'm not familiar with those functions. Can you, please, describe `U8X8_MSG_BYTE_SEND` part using words or abstracted code (like "here we transfer X bytes from buf")?

**Comment by olikraus** [comment]
> I'm not familiar with those functions. Can you, please, describe U8X8_MSG_BYTE_SEND part using words or abstracted code (like "here we transfer X bytes from buf")?

with all arg_int bytes located at arg_ptr
  prepend the byte with the last value of DC (extending the 8 bits to 9 bits)
  transmit 9 bits via SPI

**Comment by v1993** [comment]
Should I prepend each transmitted byte? That would be actually very complicated to implement (it is hard to make buffer where I store bits and not bytes AFAIK). As an alternative, I could try making transaction for each byte + bit, but that would end up in huge speed decrease.

So I'm not going to make 3-wire HW SPI eiter now or in near future I guess. Bits outside of bytes are hard to work with.

**Comment by olikraus** [comment]
> Should I prepend each transmitted byte? That would be actually very complicated to implement (it is hard to make buffer where I store bits and not bytes AFAIK). As an alternative, I could try making transaction for each byte + bit, but that would end up in huge speed decrease.

I have not invented this. It is as it is. You can read about this for example here in secton 8.1.4 (page 18): https://cdn-shop.adafruit.com/datasheets/SSD1306.pdf

> So I'm not going to make 3-wire HW SPI eiter now or in near future I guess. Bits outside of bytes are hard to work with.

This is why there is no HW 3-wire SPI for u8g2 at the moment. AVR uC can not transmit 9 bit data items.

**Comment by v1993** [comment]
Well, ESP-IDF _is_ able to transmit non-bytes. But, as I said, putting stuff into buffer would be painful and not doing so would be still (but not that much) painful, slow and probably incorrect.

**Comment by v1993** [comment]
New codebuild development is going really well! I've finished `U8g2lib.h` for arduino generation and finishing `U8x8lib.h` now.

**Comment by v1993** [comment]
@olikraus Hi! I've completed Arduino code generator part. Can you, please, try `csrc` and `cppsrc` from my repository (they are assembled by newcodebuild) and say do they work fine?

Thanks a lot!

**Comment by v1993** [comment]
Oops, sorry, forgot to push last commit when sending previous reply :)

Link to my fork is https://github.com/v1993/u8g2.

**Comment by v1993** [comment]
Well, there's a bug in C++ part which lead to generation of dublicated names. I'll investigate it.

**Comment by v1993** [comment]
That bug is fixed now. I'm able to build and run basic arduino example. :-P

P.S.: I forgot to update "6800" to "8080" when copy-pasing stuff. Also made a minor mistake in template file (forgot to remove one line).

**Comment by v1993** [comment]
Good news! Got code generation working for ESP-32.

Now I want to ask you about docs:

1. Where can I find
```
u8g2setupc.md
u8g2setupcpp.md
u8x8setupc.md
u8x8setupcpp.md
```
you mentioned before?
2. There are two different C++ interfaces now. They may be similiar, but they are different. What do you think about ESP32 version docs? Do we need to auto-generate them too? Do we need to auto-generate huge platform-specific docs in first place?

P.S.: I've implemented destructors support in last commit. You can see what I mean there: https://github.com/v1993/u8g2/commit/c253a419f37fabf6477512b7a2973577006d031a

**Comment by v1993** [comment]
As a sidenote, got 993FPS (!!!!) by replacing text drawing with box drawing. So bottleneck is frame calculation and not SPI (which is working at 80000kHz).

Also, last commit change a little bit of things on C side. You may want to check it out: https://github.com/v1993/u8g2/commit/09b645414cc7d43d35e3e07ba5bbe5514416dae7

**Comment by olikraus** [comment]
The .md files are part of u8g2.wiki.

> There are two different C++ interfaces now. They may be similiar, but they are different. What do you think about ESP32 version docs? Do we need to auto-generate them too? Do we need to auto-generate huge platform-specific docs in first place? They could be made smaller by just listing contoller/display pairs and describing naming shcemes for each applicaton (C/Arduino/ESP-IDF) with examples.

Still I do not know the benefit for u8g2 project. Do you expect me to switch to a new build environment?

**Comment by v1993** [comment]
@olikraus That would be somewhat really noice… ESP-IDF and Arduino C++ parts share a lot of code, and **ESP-IDF C++ files are only updated by new build system**. I would strongly recommend to use it.

If you worry really much, you can tweak old system a bit to match changed (a bit) comments and keep it as backup one for WSHTF scenario, so if new one fails for some reason you'll be able to at least keep arduino and docs generation working. But  using new one for regular builds is what I expect from you :-)

> Still I do not know the benefit for u8g2 project.

Ability to natively run on ESP-IDF platform with all C++ features I guess. There are many developers who prefer to write witthout Arduino HAL.

> The .md files are part of u8g2.wiki.

How can I download that docs? Where should I "install" preprocessed results?

**Comment by v1993** [comment]
@olikraus Sorry to annoy you a bit, but I need your answer to do any futher development.

**Comment by olikraus** [comment]
> @olikraus Sorry to annoy you a bit, but I need your answer to do any futher development.

Did I ever mention that my spare time is limited by job and family ;-) 
Each wiki has a clone this wiki box, so in this case, just clone "https://github.com/olikraus/u8g2.wiki.git"
Also google would have provided you this information very quickly within seconds.

**Comment by v1993** [comment]
> Did I ever mention that my spare time is limited by job and family ;-)

You did not AFAIK :D. Have similiar problems myself tho…

While this question was important, I would like to know answers to other ones I've asked.

**Comment by v1993** [comment]
In last commit (https://github.com/v1993/u8g2/commit/8a2f8c9d82bd16adeff61cf62269bd82ca0dee15) I've disabled generation of page buffers for cases when they are bigger or eqal to full framebuffer mode. While this is totally reasonable (who need page buffer when they can use full one with same or even *smaller* memory usage), it may break compatiblity just a bit. Changes, however, would be trivial: switch to full framebuffer constructor/function.

Do you approve this? Btw, changing this aspect is extremely easy now.

**Comment by v1993** [comment]
What is correct name for `u8x8_d_ssd1305_128x32_noname`-like functions? Just want to have meaningful name for args/vars/functions. :-)

**Comment by olikraus** [comment]
Sorry for the late answer. It may get even more worse in the upcoming day. I may only have smartphone access to github issue tracker for some days.

> I would like to know answers to other ones I've asked.

Oh, i reread the previous posts, but did not found a specific question. Can you repeat the same?

> Do you approve this? Btw, changing this aspect is extremely easy now.

I do not know a single case where the page buffer is bigger then the frame buffer.



> What is correct name for `u8x8_d_ssd1305_128x32_noname`-like functions? Just want to have meaningful name for args/vars/functions. :-)

I usually call these u8x8_d_xxx functions as device function. What do you mean by "args/vars/functions"?

**Comment by v1993** [comment]
> I do not know a single case where the page buffer is bigger then the frame buffer.

https://github.com/olikraus/u8g2/wiki/u8g2setupc#max7219-32x8 is a good example. There are also ones which have the same size as full one. Removing eqal ones would work as hint to developer that they can use full mode with same memory usage.

> I usually call these u8x8_d_xxx functions as device function. What do you mean by "args/vars/functions"?

I've already called them somehow, don't remember how exactly. Under args/vars/functions I mean ones in code/docs generation.

> Oh, i reread the previous posts, but did not found a specific question. Can you repeat the same?

Did the same. Turns out I actually removed them at some point of editing reply before posting… I've solved them in temporary way and plan to discuss later when making PR.

I also think that generating big portions of constructor references is bad idea especially when there are two different C++ constructors sets. I plan to generate table like this:

| Controller/Display pair | `1` buffer | `2` buffer | `3` buffer | SPI | [other interfaces will be here in columns] | UART |
|---|:---:|---|---|---|---|---|
|SSD1306_128X32_NONAME| &#9744; |&#9744;|&#9744;|&#9745;||&#9744;|

This allows having one table for C and both C++ interfaces. As for constructor reference, I want to put it like this:

> The communication protocol is part of the constructor name (see above). They differ for Arduino and ESP-IDF.
The following communication protocols are supported on Arduino:

| Communication | Description | Constructor |
|---|---|---|
| `4W_SW_SPI` | 4-wire (clock, data, cs and dc) software emulated SPI | `U8G2_X_X_4W_SW_SPI(args_here)` |

What do you think about it? I'll continue working in this direction until you answer, so that would be sad if I'll have to throw out just written piece of code :-P

**Comment by olikraus** [comment]
No doubt, this looks better than the current wiki :-)

> https://github.com/olikraus/u8g2/wiki/u8g2setupc#max7219-32x8 is a good example. There are also ones which have the same size as full one. Removing eqal ones would work as hint to developer that they can use full mode with same memory usage.

ok....

**Comment by olikraus** [comment]
Ok, I said it looks better. However i think the constructor name should be mentioned.
Additionally, the checkboxes imply, that the 1, 2 and f constructors might not be there. However, they are always created (even in those cases, where they are useless). So having checkboxes is a little bit useless.

**Comment by v1993** [comment]
> Additionally, the checkboxes imply, that the 1, 2 and f constructors might not be there. However, they are always created (even in those cases, where they are useless). So having checkboxes is a little bit useless.

I partially agree with you.

At first, constructors creation/docs generation are now united: if there is constructor, there will be docs and vice versa. But you make a good point for `F` constructor because it is always here.

> However i think the constructor name should be mentioned.

I'll think about it… don't think that listing all constructors like now is required, but giving good examples and description is important.

BTW `u8x8setupc.md` looks kinda lonely…

**Comment by olikraus** [comment]
> BTW u8x8setupc.md looks kinda lonely…

Not sure what you mean. 

In general the generated docs are a little bit less frequently used. Most users just see the constructor list in the examples :-(

I really like the effort, you put into this topic. Hope it is worth to do all this.

**Comment by v1993** [comment]
u8x8setupc.md contain only generated list and no hints at all.

> I really like the effort, you put into this topic. Hope it is worth to do all this.

Thank you! I really want to make ESP-IDF support feel complete: have C++ bindings (done) and nice docs (WIP). To keep thing in one place I do the same to Arduino part obviously (united docs/C++ generator). Hope this will be useful for other ESP-IDF community members too.

**Comment by v1993** [comment]
That's some progress on docs: https://github.com/v1993/u8g2/blob/master/tools/newcodebuild/docs/u8g2setupcpp.md. What do you think?

**Comment by olikraus** [comment]
Looks good. Still a list of all constructors would be good.

**Comment by v1993** [comment]
Well, I personally think that list of all C++ constructors is pretty bad idea: having only Arduino one seems kind of unfair to ESP-IDF users and having both would result in giant page.

**Comment by v1993** [comment]
Well, I personally think that list of all C++ constructors is pretty bad idea: having only Arduino one seems kind of unfair to ESP-IDF users and having both would result in giant page.

**Comment by v1993** [comment]
Did my reply got dublicated?

**Comment by v1993** [comment]
Wow! Thanks to table I've just fixed generation for I²C constructors.

Turns out it is really helpful tool even for development stage.

**Comment by v1993** [comment]
@olikraus I've noticed that U8x8 C++ Setup and U8g2 C++ Setup are extremely close to each other. Do you think that it is good idea to merge them into one file?

I'll do so myself.

Also, how can I send you non-generated changed/added docs from wiki?

**Comment by olikraus** [comment]
U8g2 and u8x8 are split to allow projects with pure u8x8 code

**Comment by v1993** [comment]
> U8g2 and u8x8 are split to allow projects with pure u8x8 code.

I was talking about doc pages there…

> What do you mean by send code? My email is mentioned in all c codes.

Well, I know that. For code part I can make pull request via GitHub, but what about docs? What would be a convinient way to review and merge wiki changes for you?

**Comment by v1993** [comment]
@olikraus If it isn't clear, I'm waiting for response.

**Comment by v1993** [comment]
@olikraus Umm, sorry to annoy you a bit more again… just saw that you're active in other issues. I'm waiting for your help here too. I'll have less free time in next few weeks, so I would be happy to finish docs as soon as possible (also complete notes while I do remember some details about code) and merge pull request.

**Comment by olikraus** [comment]
Sorry, as mentioned before, I am away from all kinds of computer hardware. Only mobile phone is there.  I will be absent for another 7 days 

Regarding wiki pull requests. I'd not know how to do this. But: is it required? Docs should be generated and not added via pull requests.

**Comment by v1993** [comment]
> Regarding wiki pull requests. I'd not know how to do this. But: is it required? Docs should be generated and not added via pull requests.

Yep, but… I'm going to change non-generated docs (and write some new from scratch). What about those?

Again, do you thing that it is a good idea to merge u8g2 c++ setup and u8x8 c++ setup because they are very close and basicly dublicate each other in most places?

**Comment by olikraus** [comment]
> What about those?

I could give you access rights.

u8g2 c++ setup and u8x8 c++ setup should be kept separatly.

**Comment by v1993** [comment]
> I could give you access rights.

Would it be ok to create some docs before pull reuest?

> u8g2 c++ setup and u8x8 c++ setup should be kept separatly.

Why? If I was unclear once again, I'm talking about their *docs*, not *headers*. I said that in previous reply.

**Comment by olikraus** [comment]
If someone wants to work with u8x8 only, then why shell that user read about u8g2? I think it is better to have two different setup docs.

**Comment by v1993** [comment]
@olikraus Eh, u8g2 setup and u8x8 setup (references) are basicly dublicates (they literally contain tons of eqal text) with very minor changes. Merging them is rather trivial and removes requiment to support two copies of docs.

**Comment by v1993** [comment]
@olikraus Oh, and before I forget: ESP-IDF CMakeLists.txt include `csrc/` file list.

Do you think that it is a good idea to auto-generate it too? Can I assume any platform (linux/windows/etc.) in such case?

**Comment by olikraus** [comment]
How can I answer this? 
My fear is, that with all your modifications, my projects becomes more difficult to maintain. Or I need to learn a lot of new things to maintain it. Please answer by your self and let yourself guide by the goal not to modify the existing structure.

**Comment by v1993** [comment]
> My fear is, that with all your modifications, my projects becomes more difficult to maintain.

Why? There is no way merging two docs which repeat each other a lot may hurt development!

> Or I need to learn a lot of new things to maintain it.

Well, it may apply to new codebuild, but not to docs. In fact, lua it rather easy to understand in most casesand I'll be glad to help with it.

> Please answer by your self and let yourself guide by the goal not to modify the existing structure.

So again, what is a problem of merging two pretty much dublicated docs?

Also, what about my last reply? You repo contain `.sh` scripts so I suppose it is fine to assume linux?

**Comment by olikraus** [comment]
> So again, what is a problem of merging two pretty much dublicated docs?

I just think it is better to just get exactly the information which you want to see.

> Also, what about my last reply? You repo contain .sh scripts so I suppose it is fine to assume linux?

Yes, I do all the development with Linux

**Comment by v1993** [comment]
> I just think it is better to just get exactly the information which you want to see.

Oh, I see. I'll still try to megre them, but will keep what you said in mind.

> Yes, I do all the development with Linux

Great! Means that I can freely run `find` :-)

**Comment by olikraus** [comment]
> > Yes, I do all the development with Linux
> 
> Great! Means that I can freely run `find` :-)

Sure all the usual commands are there...

**Comment by v1993** [comment]
@olikraus I can't merge your last changes because of C++ part. Can you, please, tell what was changed in generation mechanism what I have to incorportate?

New displays/controllers, different functions, etc.?

**Comment by olikraus** [comment]
I have added a new display.
